src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -173,3 +175,97 @@ g.test('readwrite_upon_aspects')
       encoder.finish();
     }, !success);
   });
+
+g.test('binding_types_combination')
```

Comment:
Maybe this should be folded in the test above and do the following combinations:

 - For each pair of usages:
 - For each combination of subresources (same, different mip, different layer, single subresource+entire conflicting layer, single subresource + entire conflicting mip)

Then the test body would get metadata on each usage (readonly vs. not) and decide whether success should be true.

Also we should be ready to factor out some of the logic to common helper.

Reply:
Corentin, My original proposal about the test plan was something like this: a super big combinations with every possible factors: different pairs of read/write bindings like the new added test, different combinations of subresources like the tests we have already added, different shader stages and visibility (another new patch), different sync scope (inside a draw calls, different draw calls but inside a render pass, different passes), etc. And combine all of these factor's together. That original test plan was really all-encompassing like your suggestion here and went even further. But it seems that is not necessary and the logic might be complex and hard to understand. So Kai and I simplified the combinations.   

However, single subresource + entire conflicting layers, single subresource + entire conflicting mip might be good to being added into the tests. And maybe we can add a few more like single subresource + entire resource (all layers, all mip levels), a single layer + entire resource, a single mip chain + entire resource. You see, it's not easy to draw a line and say "that's enough". 

WDYT, @kainino0x and @austinEng?

Reply:
I do think it's likely easier to read and write with the tests split this way. But it depends on whether we think it's necessary to test the combinations that we're missing here. I am having a hard time figuring out exactly what we're missing / deciding whether we need it.

The main thing we need to ensure is that the subresource usage tracking is working properly. That means, in each place that a subresource usage has to be tracked, we need to make sure (1) it is tracked (2) the subresource "address" (level, layer, aspect) and usage (at least, whether it's R, W, or RW) is passed into the tracking correctly.

Perhaps we don't need to cover the full combination of (different ways to read and/or write a subresource) x (different ways to address disjoint/non-disjoint ranges of subresources). But it would be good to test at least (different ways to write a subresource) x (different ways to address different subresources). And by the time you reach that it might not be hard to test more.

There's also the issue of depth/stencil render targets, and aspects. But looking at the code you have already written now, it's really not that different from the others.

I think it's possible to get complete testing here while maintaining a lot of the straightforwardness that we have now. The hard parts will probably be (1) expressing the combinatorics (maybe can be expressed as a sum of one combination for color + one combination for depth/stencil) and (2) keeping the success expectation computation simple (I think this can be done with good helper tables and methods, e.g. a helper to compute whether two subresource ranges overlap, a helper to compute whether two usages conflict, and some data in tables to tell us what binding types read, write, or read-write).

Reply:
As Corentin and Kai suggestion, I added a combination of subresources for color (same, different mip levels, different array layers, single subresource + entire mip chain, single subresource + entire array layer) and pairs of binding types, and another combination of subresources for  aspects and pairs of binding types. Note that for aspect combination, because depth/stencil texture formats can't be used as storage bindings, so the pairs of binding types are only sampled-texture and render-target.